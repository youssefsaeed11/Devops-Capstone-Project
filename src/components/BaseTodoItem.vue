<template>
  <div
    class="BaseTodoItem"
    :class="{ 'BaseTodoItem--completed': item.completedAt }"
  >
    <div class="BaseTodoItem__checkbox">
      <slot name="checkbox">
        <input
          type="checkbox"
          disabled
          :checked="!!item.completedAt"
        >
      </slot>
    </div>
    <div class="BaseTodoItem__content">
      <slot name="content">
        {{ item.content }}
      </slot>
    </div>
    <div class="BaseTodoItem__actions">
      <slot name="actions" />
    </div>
  </div>
</template>

<script>
/**
 * @module BaseTodoItem
 */
export default {
  name: 'BaseTodoItem',
  props: {
    item: {
      type: Object,
      required: true
    }
  }
}
</script>
<style lang="scss" scoped>
.BaseTodoItem {
  display: flex;
  align-items: center;
  margin-bottom: $theme--spacing-m;
  background: $theme--color-white;
  padding: $theme--spacing-s;
  border-radius: 5px;

  &--completed {
    text-decoration: line-through;
    opacity: .5;
  }

  &__content {
    flex: 1 1 auto;
    padding: 0 $theme--spacing-m;
  }

  &__actions {
    display: flex;
    align-items: center;
    justify-content: center;
  }
}
</style>
