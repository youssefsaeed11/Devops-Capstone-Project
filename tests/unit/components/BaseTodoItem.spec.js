import BaseTodoItem from '@/components/BaseTodoItem'
import { mount } from '@vue/test-utils'
import { notDeletedItem, completedItem } from '../data/ToDoFactory'

/**
 * @param overrides
 * @return {{wrapper: Wrapper<Vue>, config: any}}
 */
const createWrapper = (overrides) => {
  let config = Object.assign(
    {
      propsData: {
        item: notDeletedItem()
      }
    },
    overrides
  )

  const wrapper = mount(BaseTodoItem, config)

  return {
    config,
    wrapper
  }
}

describe('BaseTodoItem', () => {
  it('renders the BaseTodoItem', () => {
    const { wrapper, config } = createWrapper()
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.text()).toContain(config.propsData.item.content)
  })

  it('sets the completed class if completed', () => {
    const item = completedItem()
    const { wrapper } = createWrapper({
      propsData: {
        item
      }
    })
    expect(wrapper.classes()).toContain('BaseTodoItem--completed')
  })

  it('displays a disabled checkbox in the "checkbox" slot by default', () => {
    const item = completedItem()
    const { wrapper } = createWrapper({
      propsData: {
        item
      }
    })
    const checkbox = wrapper.find('input')
    expect(checkbox.exists()).toBeTruthy()
    expect(checkbox.attributes('disabled')).toBeTruthy()
    expect(checkbox.element.checked).toBe(true)
  })

  it('allows replacing the default checkbox', () => {
    const { wrapper } = createWrapper({
      propsData: {
        item: notDeletedItem()
      },
      slots: {
        checkbox: '<div class="someDiv"/>'
      }
    })
    expect(wrapper.find('.someDiv').exists()).toBe(true)
    expect(wrapper.find('input').exists()).toBe(false)
  })

  it('allows replacing the content slot', () => {
    const { wrapper } = createWrapper({
      propsData: {
        item: notDeletedItem()
      },
      slots: {
        content: '<div class="someContent"/>'
      }
    })
    expect(wrapper.find('.someContent').exists()).toBe(true)
  })
})
