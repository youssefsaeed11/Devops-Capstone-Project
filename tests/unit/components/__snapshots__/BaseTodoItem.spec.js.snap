// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseTodoItem renders the BaseTodoItem 1`] = `
<div class="BaseTodoItem">
  <div class="BaseTodoItem__checkbox"><input type="checkbox" disabled="disabled"></div>
  <div class="BaseTodoItem__content">
    notDeleted
  </div>
  <div class="BaseTodoItem__actions"></div>
</div>
`;
